api/src/main/java/org/apache/iceberg/types/TypeUtil.java

```diff
@@ -181,11 +181,34 @@ public static Map<Integer, String> indexQuotedNameById(
     return indexer.byId();
   }
 
+  /**
+   * Creates a mapping from lower-case field names to their corresponding field IDs.
+   *
+   * <p>This method iterates over the fields of the provided struct and maps each field's name
+   * (converted to lower-case) to its ID. If two fields have the same lower-case name, an
+   * `IllegalArgumentException` is thrown.
+   *
+   * @param struct the struct type whose fields are to be indexed
+   * @return a map where the keys are lower-case field names and the values are field IDs
+   * @throws IllegalArgumentException if two fields have the same lower-case name
```

Comment:
Minor: Should this be `IllegalArgumentException`? I think `IllegalStateException` may be more accurate because the state of the partition spec can't produce a case insensitive index.

Reply:
I, too, think IllegalStateException is more semantically correct, but this suggestion conflicts with your other suggestion to use Preconditions.checkArgument which throws IllegalArgumentException. Which solution has higher precedence, throwing IllegalStateException or calling Preconditions.checkArgument?